juju/model.py

```diff
@@ -0,0 +1,172 @@
+            self[key] = entity
+        return entity
+
+
+class Application:
+    def __init__(self, name, entity_cache):
+        self.name = name
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class Unit:
+    def __init__(self, name, entity_cache):
+        self.name = name
+        self.app = entity_cache.get(Application, name.split('/')[0])
+
+    def __repr__(self):
+        return f'<{type(self).__module__}.{type(self).__name__} {self.name}>'
+
+
+class LazyMapping(Mapping, ABC):
+    @abstractmethod
+    def _load(self):
+        raise NotImplementedError()
+
+    @property
+    def _data(self):
+        if not hasattr(self, '_lazy_data'):
+            self._lazy_data = self._load()
+        return self._lazy_data
+
+    def __contains__(self, key):
+        return key in self._data
+
+    def __len__(self):
+        return len(self._data)
+
+    def __iter__(self):
+        return iter(self._data)
+
+    def __getitem__(self, key):
+        return self._data[key]
+
+
+class RelationMap(LazyMapping):
+    """Map of relation names to lists of Relation instances."""
+    def __init__(self, relation_names, backend, entity_cache):
+        self._relation_names = relation_names
+        self._backend = backend
+        self._entity_cache = entity_cache
+
+    def _load(self):
+        data = {}
+        for relation_name in self._relation_names:
+            relations = data[relation_name] = []
+            for relation_id in self._backend.relation_ids(relation_name):
+                relations.append(Relation(relation_name, relation_id, self._backend, self._entity_cache))
+        return data
+
+
+class Relation:
+    def __init__(self, relation_name, relation_id, backend, entity_cache):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self.data = RelationData(self.relation_name, relation_id, self._backend, entity_cache)
+
+    @property
+    def apps(self):
+        return [unit.app for unit in self.units]
+
+    @property
+    def units(self):
+        return list(self.data.keys())
+
+
+class RelationData(LazyMapping):
+    def __init__(self, relation_name, relation_id, backend, entity_cache):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend
+        self._entity_cache = entity_cache
+
+    def _load(self):
+        data = {}
+        for unit_name in self._backend.relation_list(self.relation_id):
+            unit = self._entity_cache.get(Unit, unit_name)
+            data[unit] = RelationDataBag(self.relation_id, unit, self._backend)
+        # Juju's relation-list doesn't include the local unit(s), even though they are part of
+        # the relation. Technically, you can also call relation-get for your peers' data, but
+        # we don't want to support that, so we only manually add this local unit.
```

Comment:
This was about the fact that, from `myapp/0`, you can technically do `relation-get -r $rel_id - myapp/1` and see what data your peer has presented on the relation (for any relation, not just peer relations).  Typically, only the local unit is used in this way, from my experience. My assumption is that we don't want to surface our peer units in the relation data for arbitrary relations in the Python layer to discourage anyone from starting to rely on it.  It would also be challenging to implement technically because we don't have a reliable way to get the list of peer units without either an implicit peer relation or using `goal-state`.

The comment has been removed in the updated PR, but if my assumption about us not wanting to support / surface our peer units in arbitrary relation data is wrong, then we should discuss it further.